ngraph/core/src/op/einsum.cpp

```diff
@@ -0,0 +1,277 @@
+// Copyright (C) 2018-2021 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include <algorithm>
+#include <cctype>
+#include <ngraph/validation_util.hpp>
+#include <regex>
+#include <string>
+#include <unordered_map>
+
+#include "itt.hpp"
+#include "ngraph/op/einsum.hpp"
+
+using namespace std;
+using namespace ngraph;
+
+NGRAPH_RTTI_DEFINITION(op::v7::Einsum, "Einsum", 7);
+
+op::v7::Einsum::Einsum(const OutputVector& inputs, const std::string& equation)
+    : Op(inputs)
+    , m_equation(equation)
+{
+    constructor_validate_and_infer_types();
+}
+
+bool op::v7::Einsum::parse_equation(std::vector<std::string>& input_subscripts,
+                                    std::string& output_subscript)
+{
+    // remove extra white-spaces from the equation
+    m_equation.erase(std::remove_if(m_equation.begin(), m_equation.end(), ::isspace),
+                     m_equation.end());
+
+    // split equation to input subscripts and an output subscript
+    auto pos_output_delimeter = m_equation.find("->");
+    auto input_subscripts_str = m_equation.substr(0, pos_output_delimeter);
+
+    // split the input subscripts into a vector of input subscripts
+    input_subscripts.clear();
+    std::istringstream input;
+    input.str(input_subscripts_str);
+    for (std::string input_subscript; std::getline(input, input_subscript, ',');)
```

Comment:
I think you should fuse three loops in one

Reply:
Fixed.